### ISSUE TITLE
Line Ending Changing with formatting

### ISSUE THREAD

**Issue by DragonCrafted87** [executablebooks/mdformat#224]
**Describe the bug**
File is always output with crlf line endings on windows even if no formatting was changed

**To Reproduce**

1. Start with a valid markdown file on Windows with lf(unix) line endings
2. run file through mdformat
3. file now has crlf(dos) line endings

**Expected behavior**
File has same line ending style it started with

**Environment**

Windows 10
Python 3.9.5
pre-commit 2.13.0
-   repo: https://github.com/executablebooks/mdformat
    rev: 0.7.6
    hooks:
    -   id: mdformat
        additional_dependencies:
        -   mdformat-gfm
        -   mdformat-black

**Comment by hukkin** [comment]
Thanks for the issue!

To clarify, mdformat never creates `crlf`. It always normalizes end of lines to a single `lf` character. This is part of the whitespace normalization that mdformat does, and is by design.

If there's a good reason why one would want to use `crlf` instead, we can consider adding a `--crlf` flag. I'm not aware of a good use case, and allowing `crlf` adds some complexity (also in plugins etc.), so I haven't considered adding this.

Based on what you wrote, it seems you too may be happier with `lf`s than `crlf`s?

**Comment by DragonCrafted87** [comment]
yeah, I do want `lf` endings which is why I created the bug I've created a repo with a minimal pre-commit setup that demonstrates what I mean https://github.com/DragonCrafted87/temp-test-repo.

I have the pre-commit-hooks mixed-line-ending hook set to make sure everything is `lf` followed by mdformat which is writing out changes and then if I run them again with `pre-commit run --all-files` both show changes again.

in the case of the repo I created I would expect no changes on either run as all I've done to the repo was had GitHub initializes it and add the pre-commit hook setup yet on the 2nd and following runs I get this for output:

Mixed line ending........................................................Failed
- hook id: mixed-line-ending
- exit code: 1
- files were modified by this hook

README.md: fixed mixed line endings

Check for added large files..............................................Passed
mdformat.................................................................Failed
- hook id: mdformat
- files were modified by this hook

**Comment by hukkin** [comment]
I'm unable to reproduce on Linux:
```console
foo@bar:~$ git clone git@github.com:DragonCrafted87/temp-test-repo.git
Cloning into 'temp-test-repo'...
remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 7 (delta 0), reused 3 (delta 0), pack-reused 0
Receiving objects: 100% (7/7), done.
foo@bar:~$ cd temp-test-repo/
foo@bar:~/temp-test-repo$ pre-commit run -a
[INFO] Initializing environment for https://github.com/executablebooks/mdformat:mdformat-black,mdformat-gfm.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/executablebooks/mdformat.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Mixed line ending........................................................Passed
Check for added large files..............................................Passed
mdformat.................................................................Failed
- hook id: mdformat
- files were modified by this hook
foo@bar:~/temp-test-repo$ pre-commit run -a
Mixed line ending........................................................Passed
Check for added large files..............................................Passed
mdformat.................................................................Passed
foo@bar:~/temp-test-repo$ pre-commit run -a
Mixed line ending........................................................Passed
Check for added large files..............................................Passed
mdformat.................................................................Passed
foo@bar:~/temp-test-repo$
```

My hunch says the problem is with the `mixed-line-ending` hook you're using. The source for it is [here](https://github.com/pre-commit/pre-commit-hooks/blob/0bedbb96b93e08455ab52943c653669ed80bc356/pre_commit_hooks/mixed_line_ending.py). It uses `open(filename)` to read the file but does not specify `encoding="utf-8"`. The default encoding on Windows is not UTF-8 so this could result in the types of problems you're having.

**Comment by DragonCrafted87** [comment]
## ByteCode Values
Starting Hex View from my editor with file set to unix line endings
00000000h: 23 20 74 65 6D 70 2D 74 65 73 74 2D 72 65 70 6F ; # temp-test-repo
00000010h: 0A                                              ; .

After running just `mixed-line-ending`
00000000h: 23 20 74 65 6D 70 2D 74 65 73 74 2D 72 65 70 6F ; # temp-test-repo
00000010h: 0A                                              ; .

after running just `mdformat`
00000000h: 23 20 74 65 6D 70 2D 74 65 73 74 2D 72 65 70 6F ; # temp-test-repo
00000010h: 0D 0A                                           ; ..

## Probable cause of Issue
I did some searching and found this [stackoverflow](https://stackoverflow.com/questions/9184107/how-can-i-force-pythons-file-write-to-use-the-same-newline-format-in-windows) that lead me to look at pythons open function [documentation](https://docs.python.org/3.8/library/functions.html#open) the relevant section 

> newline controls how universal newlines mode works (it only applies to text mode). It can be None, '', '\n', '\r', and '\r\n'. It works as follows:
>
> - When reading input from the stream, if newline is None, universal newlines mode is enabled. Lines in the input can end in '\n', '\r', or '\r\n', and these are translated into '\n' before being returned to the caller. If it is '', universal newlines mode is enabled, but line endings are returned to the caller untranslated. If it has any of the other legal values, input lines are only terminated by the given string, and the line ending is returned to the caller untranslated.
>
> - When writing output to the stream, if newline is None, any '\n' characters written are translated to the system default line separator, os.linesep. If newline is '' or '\n', no translation takes place. If newline is any of the other legal values, any '\n' characters written are translated to the given string.

So it looks like something similar is likely to happen on mac as well since without a newline being set on the output file open it uses the system default.

## Proposed Solution
add a new --line-ending argument that defaults to the current behavior (b/c there are probably people relying on that with knowing about it) that takes in one of lf, cr, or crlf and sets output file accordingly.

assuming that solution is acceptable I should be able to find time over the weekend the generate a PR for the implementation therein

**Comment by hukkin** [comment]
This should be fixed in v0.7.7
